rfd/0178-github-proxy.md

```diff
@@ -0,0 +1,853 @@
+- Alice can still logs into GitHub through a browser and make chnages there if
+   necessary.
+- Alice can manually sign an user certificate according to [GitHub
+  spec](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-git-access-to-your-organizations-repositories/about-ssh-certificate-authorities#issuing-certificates).
+  The CA used to generate the user certificate could be a backup of Teleport's
+  CA that is exported to a 3rd party secret store. Or, Alice can generate a
+  self-signed CA on the spot and import it to the Organization settings
+  temporarily.
+
+In addition, trusted GitHub environments like GitHub actions are not affected
+by the option to disallow personal access tokens. GitHub allows these services
+to continue use existing authentication methods so these services do not need
+to go through Teleport.
+
+### Implementation
+
+#### Overview
+```mermaid                                
+sequenceDiagram
+    participant git
+    participant tsh
+    participant client browser
+    participant Proxy
+    participant Auth
+    participant GitHub
+                                
+    git->>tsh: sshCommand="tsh git ssh"
+    alt no GitHub user ID
+      tsh->>Proxy: CreateGitHubAuthRequestForUser
+      Proxy->>Auth: CreateGitHubAuthRequestForUser
+      tsh->> client browser: open
+      client browser->> GitHub: redirect
+      GitHub<<->>Proxy: callback
+      Proxy<<->>Auth: verify callback and generate new cert with GitHub user ID
+      client browser->>tsh: new cert with GitHub user ID
+    end
+    tsh->>Proxy: SSH transport and RBAC
+    Proxy->>Auth: GenerateGitHubUserCert
+    Auth->>Proxy: signed cert with "id@github.com" ext
+    Proxy->>GitHub: forward SSH
+    git <<->>GitHub: git pack-protocol
+```
+
+#### The GitHub OAuth flow
+
+Teleport needs user's GitHub ID or username in order to sign an user
+certificate to authenticate with GitHub.
+
+One option to achieve this is to allow the administrator to input each user's
+GitHub identity information as a trait. However, this was deemed insecure, as
+an admin could potentially alter a user's GitHub identity to impersonate
+another GitHub user.
+
+Instead, we use the GitHub OAuth flow to retrieve each user's GitHub identity.
+This process is similar to the existing GitHub SSO flow; however, while GitHub
+SSO creates a new user, this GitHub OAuth flow is designed to update the login
+state for an already authenticated user.
+
+To be more specific, the current GitHub SSO flow initiates an unauthenticated
+HTTP call to Teleport to create a GitHub auth request. Once the Auth service
+verifies the GitHub callback, it creates a new user associated with a GitHub
+connector.
+
+In contrast, the new flow uses an authenticated gRPC call to Teleport to create
+the GitHub auth request. Once the Auth service verifies the GitHub callback, it
+attaches the GitHub identity information directly to the authenticated user who
+initiated the request.
```

Comment:
To follow up on my question above, sounds like this should work for scenarios where you do `tsh login` from a development box and not your laptop as well, correct? Have you tested this scenario?

Reply:
As discussed in person, one mitigation we decided to do is to preserve github identity info so user only needs to do this **once** on their laptop.

I found today user login state is refreshed upon new user login. I have updated the refresh logic to preserve the github identity info during refresh so that another login wouldn't void the OAuth flow. We might need a way to properly void the GitHub identity saved though. I might add a tctl call later.

The other alternative is port-forwarding. For example, from user's laptop:
```
(local)$ tsh ssh -L 8080:localhost:8080 bob@remote-box
(remote)$ tsh login xxxx
(remote)$ tsh git login --github-org my-org --callback-addr localhost:8080
```
Then `bob` can open the URL on his laptop's browser to finish the flow.

I will update the RFD to reflect this.